Reimplement #1652 - switchable time formats

daschuer opened the pull request:
This is a rebased Version of #1915 onto the original #1652 
to keep the original commit messages.

poelzi committed:
Implement switchable duration format

The display format on the decks can now be configured.
Implement alternative display based on kiloseconds.

poelzi committed:
Cleanup and Hectoseconds

Fix comments regarding style.
Also add hectosecond formatting. For mixing, 100 seconds is actually a quite useful
seperator.

poelzi committed:
add simple second display.

rename formatSeconds into formatTime to reflect function better.

poelzi committed:
fixup tests. Try some unrelated character as seperator in hectoseconds.

poelzi committed:
Merge remote-tracking branch 'origin/master' into feature/seconds

benis committed:
Implement switchable time format for decks by @poelzi - #1652

benis committed:
Comment to identify separator characters

benis committed:
Fix issues identified by CodeFactor

benis committed:
Replace if with VERIFY_OR_DEBUG_ASSERT in src/util/duration.cpp

daschuer committed:
Merge commit 'f325270ee31bcdad96e1f206637f059ad72e6b01' into seconds

daschuer committed:
reintroduce the lost "?" for negative durations

daschuer commented:
LGTM, waiting for CI

benis commented on src/util/duration.cpp:
```diff
@@ -25,9 +25,9 @@ QChar DurationBase::kHectosecondSeparator = QChar(0x231E);
 // static
 QString DurationBase::formatTime(double dSeconds, Precision precision) {
-    VERIFY_OR_DEBUG_ASSERT (dSeconds >= 0.0) {
+    if (dSeconds < 0.0) {
```
So should this not be using VERIFY_OR_DEBUG_ASSERT?

benis commented:
There is still some work to do on this. Is there a way for me to push to this PR?

Also, the tests you said were failing - how do you run those?

benis committed:
Add 'Traditional (Coarse)' time format option

benis committed:
Minor formatting corrections to duration.cpp

benis commented:
I would hope to have this done by the end of January for sure.

I've just created a PR to your hectoseconds branch with the change for Traditional (Coarse) time format.

benis commented:
I still need to check if the tests are working and figure out why not if not. That's the only critical thing before this could be merged.

Here are some other things I would like to look at:
* Can we find a more conventional choice of separator character?

Also, the options displayed in the Deck Preferences are a bit inconsistent:
<img width="355" alt="screenshot 2019-01-10 at 21 25 10" src="https://user-images.githubusercontent.com/36423450/50998061-3b52e000-151e-11e9-9dc3-be2d26b43eaa.png">
* is there a way to indicate that the hh is optional, or could we just remove it? Not many tracks over an hour long.
* 's zz' uses the defined centisecond separator character (thin space) - should it?
* 'k.sss zz' uses a static period, even though the actually-displayed character might be different. Then the defined centisecond separator character is used between sss and zz
* 'hs.ss zz' also has a static period but uses the hectosecond separator character between ss and zz
(See dlgprefdeck.cpp:122 onwards for relevant code.)

I also noticed that if you choose a different time format in Preferences but then hit Cancel, when you next load Preferences the drop-down has retained your value, so if you go back in to change something else, the deck preferences then get applied too.

benis committed:
Re-format dlgprefdeck.cpp:120 onwards to make more readable

benis commented:
Just ran the tests, if I've understood the output correctly they now pass.

daschuer commented:
Formating:
Let's follow strictly the SI recommendation in en-US format.
Decimal separator .
Thouthands colon , 
Decimal places seperator thin space.

Seconds ss,sss.ss
Kiloseconds ss.sss ss
Hectoseconds sss.ss ss

For the hh: I think a minute only format is also common. A movie runtime is given as 98 min for example. So we can pick IMHO that and drop the hh: entirely.

benis committed:
Remove 'hh:' from descrptions in time format drop-down in Deck Preferences

benis commented:
I don't think your examples match up with SI. The [SI standard](https://en.wikipedia.org/wiki/ISO_31-0#Numbers) says that decimal places can be separated by a point or a comma (the former in US), and that the thousand separator should be a space. That makes sense to me, could we also use it as the hundred separator in the hectoseconds format?

benis committed:
Clean up deck time formatting logic

The choice of variable names for separator characters (both 'group'
separator characters and decimal separator), as well as the actual
application of formatting, was inconsistent and confusing. Now it's
better, hopefully.

benis committed:
Fix entries in deck time prefs drop-down

These were using a mix of hard-coded and argument-based separator
characters and also didn't match the intended formatting. This is
now fixed and all separator characters are now argument-based.

benis committed:
Amend durationutiltest.cpp following previous 2 commits

benis commented:
I've pushed the last of my commits. Here's how everything looks:

Preferences
<img width="353" alt="prefs" src="https://user-images.githubusercontent.com/36423450/51077562-45045100-16a0-11e9-9a6e-d8165c729c1e.png">

Traditional
<img width="369" alt="traditional" src="https://user-images.githubusercontent.com/36423450/51077559-446bba80-16a0-11e9-9471-f1878800597d.png">

Traditional (Coarse)
<img width="362" alt="traditional coarse" src="https://user-images.githubusercontent.com/36423450/51077560-446bba80-16a0-11e9-8c12-4ed78019734a.png">

Seconds
<img width="357" alt="seconds" src="https://user-images.githubusercontent.com/36423450/51077561-446bba80-16a0-11e9-8bcc-b4e35f2f2658.png">

Kiloseconds
<img width="365" alt="kiloseconds" src="https://user-images.githubusercontent.com/36423450/51077558-446bba80-16a0-11e9-855c-58cca28c5739.png">

Hectoseconds
<img width="357" alt="hectoseconds" src="https://user-images.githubusercontent.com/36423450/51077557-446bba80-16a0-11e9-9d45-0b88897e9dc0.png">

uklotzde commented on src/preferences/dialog/dlgprefdeck.h:
```diff
@@ -20,7 +20,14 @@ namespace TrackTime {
+    };
+
+    enum class DisplayFormat {
+        TRADITIONAL,
```
Why do DisplayMode and DisplayFormat use inconsistent naming conventions for their enum literals?

benis replied:
Good question. You're referring to the fact that one is in Sentence Case and the other in SNAKE_CASE, right?

uklotzde replied:
Of course, it's obvious ;)

uklotzde commented on src/util/duration.cpp:
```diff
@@ -32,12 +38,87 @@ QString DurationBase::formatSeconds(double dSeconds, Precision precision) {
+QString DurationBase::formatSeconds(double dSeconds, Precision precision) {
+    if (dSeconds < 0.0) {
+        // negative durations are not supported
+	return "?";
```
Use a constant QString named kInvalidDurationString.

daschuer commented:
Traditional OK 
Traditional (Coarse) OK
Seconds OK 
Kiloseconds NOK  
* This is part of a kiloseconds movement https://kilosecond.info/ the 
* I think it should be 0.000 62  -1.000 95
hectoseconds I don't care much  
* I think it should be 0.00 62  -10.00 95 according to SI.

daschuer committed:
Merge remote-tracking branch 'upstream/master' into hectoseconds

benis commented:
> Traditional OK
> 
> Traditional (Coarse) OK
> 
> Seconds OK
> 
> Kiloseconds NOK
> 
>     * This is part of a kiloseconds movement https://kilosecond.info/ the
> 
>     * I think it should be 0.000 62  -1.000 95
> 
> 
> Hectoseconds I don't care much
> 
>     * I think it should be 0.00 62  -10.00 95 according to SI.

I still don't think that's consistent with SI, but feel free to change it.

benis committed:
Refactor DisplayMode enum for consistency

benis committed:
Return const QString for invalid duration value

daschuer committed:
Merge pull request #38 from benis/hectoseconds2

Additional minor fixes

benis commented:
https://en.wikipedia.org/wiki/International_System_of_Units#General_rules

* The symbol for the decimal marker is either a point or comma on the line
...
* Spaces should be used as a thousands separator (1 000 000) in contrast to commas or periods (1,000,000 or 1.000.000) to reduce confusion resulting from the variation between these forms in different countries.

So spaces for thousands, point or comma for decimal marker. You're proposing to do the opposite.

benis commented on src/util/duration.h:
```diff
@@ -14,6 +14,7 @@ namespace mixxx {
 class DurationBase {
 
   public:
+
```
I seem to have introduced this empty line in error.

daschuer commented:
> So spaces for thousands, point or comma for decimal marker. You're proposing to do the opposite.

I read this like this:
1 ks = 16 min + 40 s
1 000 ks = 11 d + 13 h + 64 min + 40 s 
0.9 ks = 15 min 
0.001 ks = 1 s
0.000 001 hs = 1 ms 

1 hs = 1 min + 40 s
1 000 hs = 27 h + 46 min + 40 s 
0.9 hs = 1.5 min 
0.001 hs = 100 ms
0.000 01 hs = 1 ms

benis commented:
Ah, I see where you're coming from now, sorry for the confusion. Are you happy to change it? It will be less faff than me submitting another PR to you.

daschuer commented:
OK I will do the change, I am only unsure about hecto seconds. 
should we go with the SI correct:
0.000 1 hs = 10 ms
with the nicer looking: 
0.00 01 hs = 10 ms

benis commented:
I vote for the second one. It's kind of an unconventional format so why not.

daschuer committed:
make time formats more like SI

daschuer commented:
OK, done. 

However, I have just played with it and it feels odd. 
I am considering to replace it with a plain seconds format with three leading zeros. 
000.00

Or should we reintroduce your original hectoseconds format? 
What do you think?

benis commented:
I'm not attached to the hectoseconds format, so I wouldn't mind if we replaced it. I think your suggestion is a better choice and more likely to get used.

daschuer commented:
OK I will change it.

daschuer committed:
Replace hectoseconds format by seconds long. Hectosends fromat was too hard to read and the new long format is basically the same but with other separators

daschuer commented:
Done. What else is missing before merge?

benis commented:
I think that's everything. I inadvertently introduced a blank line in src/util/duration.h (:17) so I guess this would be a good time to fix that, but otherwise this is good to go.

daschuer commented:
OK, I am fine with this now. 
We may consider to alter also the library column accordingly, but that is a separate PR.

@uklotzde Is anything left to do before merge?

daschuer commented:
Thank you @benis and @poelzi for working on this.

uklotzde commented:
New bug: https://bugs.launchpad.net/mixxx/+bug/1830488
